course_catalog/etl/deduplication.py

```diff
@@ -0,0 +1,113 @@
+    """
+    Helper function to determine the most relevant course run.
+
+    Args:
+        runs (QuerySet): a set of LearningResourseRun objects
+    Returns:
+        A LearningResourseRun object
+    """
+
+    # if there is a current run in the set pick it
+    most_relevant_run = next(
+        (run for run in runs if run.availability == AvailabilityType.current.value),
+        None,
+    )
+
+    if not most_relevant_run:
+        # if there a future runs in the set, pick the one with earliest start date
+        runs = runs.order_by("best_start_date")
+        most_relevant_run = next(
+            (
+                run
+                for run in runs
+                if run.availability
+                in [
+                    AvailabilityType.upcoming.value,
+                    AvailabilityType.starting_soon.value,
+                ]
+            ),
+            None,
+        )
+
+        if not most_relevant_run:
+            # get latest past run by start date
+            most_relevant_run = next((run for run in runs.reverse()))
+
+    return most_relevant_run
+
+
+def course_fields_to_copy():
+    """
+    Helper function that returns fields to copy from the newest course to the most relevant course
+    """
+
+    fields = [f.name for f in Course._meta.fields]
+
+    for field in ["id", "created_on", "updated_on", "course_id", "published"]:
+        fields.remove(field)
+
+    return fields
+
+
+def copy_course_data(first_course, most_relevant_course, courses):
+    """
+    Copy course information from the most relvant copy of the course to the earliest copy
+
+    Args:
+        runs (QuerySet): a set of LearningResourseRun objects
+    Returns:
+        A LearningResourseRun object
+    """
+
+    # copy over all course offered bys to the first course
+    offered_bys = [set(course.offered_by.all()) for course in courses]
+    offered_bys = reduce(lambda x, y: x | y, offered_bys)
+    first_course.offered_by.set(offered_bys)
+
+    for attr in course_fields_to_copy():
+        setattr(first_course, attr, getattr(most_relevant_course, attr))
+
+    first_course.published = True
+    first_course.save()
+
+
+def dedupe_courses(platform):
+    """
+    Function to combine course runs from duplicate courses and remove duplicate courses from the search
+    """
+
+    for duplicate_list in load_course_duplicates(platform):
+        courses = Course.objects.filter(course_id__in=duplicate_list)
+        runs = [course.runs.all() for course in courses]
+
+        if not runs:
+            continue
+
+        runs = reduce(lambda x, y: x | y, runs)
+
+        runs = runs.order_by("created_on")
+        first_run = next(run for run in runs)
+        runs.update(object_id=first_run.object_id)
+        most_relevant_run = get_most_relevant_run(runs)
+
+        copy_course_data(
+            courses.get(pk=first_run.object_id),
+            courses.get(pk=most_relevant_run.object_id),
+            courses,
+        )
```

Comment:
It may have slipped my mind since it's been almost a week, but could you give some context around why this copying of fields was added in your latest changes? My gut says if we're having to do something like this we maybe need to back up and look at the design here again.

Reply:
We were concerned that sharable links would no longer work if the visible copy for a duplicate course switched over to a newer copy. This keeps the original course but overwrites course content with the data corresponding to the most relevant run.

As we discussed last week, we don't know which copy of a course has the latest data until all the copies are loaded